proposals/0698-qualified-literals.rst

```diff
@@ -0,0 +1,359 @@
+
+Parser
+~~~~~~
+
+Update `Section 10.5 <https://www.haskell.org/onlinereport/haskell2010/haskellch10.html#x17-18000010.5>`_ of the Haskell 2010 report as follows.
+
+.. code-block:: abnf
+
+  aexp → qvar
+       | ...
+       | modid . integer
+       | modid . ( {-} integer )
+       | modid . ( float )
+       | modid . string
+       | modid . multiLineString
+       | modid . [ exp_1 , ..., exp_k ]
+
+  apat → var [ @ apat ]
+       | ...
+       | modid . integer
+       | modid . ( {-} integer )
+       | modid . ( float )
+       | modid . string
+       | modid . multiLineString
+       | modid . [ pat_1 , ..., pat_k ]
+       | modid . ( pat_1 : ... : pat_k )
+
+
+Proposed Library Change Specification
+-------------------------------------
+
+Template Haskell
+~~~~~~~~~~~~~~~~
+
+The following constructors will be updated to the following definitions:
+
+::
+
+  ListE (Maybe ModuleName) [Exp]
+
+  StringL (Maybe ModuleName) String
+  IntegerL (Maybe ModuleName) Integer
+  RationalL (Maybe ModuleName) Rational
+
+Examples
+--------
+
+Vector
+~~~~~~
+
+Currently, if you want to pattern match on vector, you have to use ``OverloadedLists`` (which enables it for list literals in the entire file) or be verbose:
+
+::
+
+  case user of
+    -- guard
+    User{tags = tags} | ["a", tag2] <- V.toList tags -> _
+    -- with ViewPatterns
+    User{tags = (V.toList -> ["a", tag2])} -> _
+
+With ``QualifiedLiterals``, ``vector`` could define:
+
+::
+
+  module Data.Vector.Qualified where
+
+  fromList = V.fromList
+  listCons = (:)
+  listNil = []
+
+  listUncons = V.uncons
+
+And the user could do:
+
+::
+
+  import Data.Vector.Qualified qualified as V
+
+  case user of
+    User{tags = V.["a", tag2]} -> _
+
+One scenario this can come up is when parsing ``Aeson.Array``, which stores JSON values in a ``Vector``.
+
+Scientific
+~~~~~~~~~~
+
+`Scientific <https://hackage.haskell.org/package/scientific-0.3.8.0/docs/Data-Scientific.html#t:Scientific>`_ represents an arbitrary precision number. It has a ``Num`` instance, but ``+`` and ``-`` are unsafe and can cause OOM. Safety-minded developers might desire to wrap with a newtype that provides ``unsafeAdd`` but not ``+``, to prevent call-sites from accidentally blowing up memory.
+
+::
+
+  newtype BigNumber = BigNumber Scientific
+
+  unsafeAdd :: BigNumber -> BigNumber -> BigNumber
+  unsafeAdd = coerce (+)
+
+Writing tests for functions using ``BigNumber`` is arduous, since repeating ``BigNumber 123`` is verbose. You could write a helper function ``big = BigNumber``, but this is unsafe if accidentally called on a non-literal, as ``Scientific`` throws a runtime error if converting from a repeating decimal.
+
+With QualifiedLiterals, you could write ``Big.123``, which guarantees that ``Big.fromRational`` is only called on literals (e.g. you could configure hlint to ban calling ``Big.fromRational`` directly and only be used via QualifiedLiterals).
+
+Positive-only literals
+~~~~~~~~~~~~~~~~~~~~~~
```

Comment:
The benefit in this section is from introducing a notion of a positive only literal, but afaict that is orthogonal to the current proposal.

For instance we could have a `PositiveNum` typeclass to implement this.